Run Cluster Create/Update/Delete tasks sequentially per SDDC

dimitarproynov opened the pull request:
If two concurrent cluster tasks run per SDDC we get the following error:
com.vmware.skyscraper.common.InternalSkyscraperException: Concurrent update on <SDDC-ID>, column name {1}. Try operation again.

fix for: https://github.com/vmware/terraform-provider-vmc/issues/93

Use a per SDDC lock to force the Terraform provider to wait for the previous task completion.

Now that an almost full E2E suite is present, the pattern of the copy-pasted "wait until finished or error" code across the provider's resources can be extracted, as a helper function to avoid code duplication, safely.

In addition to the refactorings a Spectral scan action has been added following a recommendation from a VMware security architect. 

Testing done:
make build
make test
golangci-lint run

Added a second cluster in the TestAccResourceVmcClusterZerocloud

ran successfully all acceptance tests:
make testacc TESTARGS="-run='TestAccResourceVmcSddcZerocloud|TestAccResourceVmcClusterZerocloud|TestAccResourceVmcSiteRecoveryZerocloud|TestAccResourceVmcSrmNodeZerocloud|TestAccDataSourceVmcCustomerSubnetsBasic|TestAccDataSourceVmcConnectedAccountsBasic|TestAccDataSourceVmcOrgBasic|TestAccDataSourceVmcSddcBasic' -parallel 4

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

dimitarproynov committed:
Merge pull request #3 from dimitarproynov/refactor_unit_tests

Refactor unit tests

dimitarproynov committed:
Remove environment

The main repo for the terraform vmc provider doesn't support the GitHub "Environments" feature.

dimitarproynov committed:
Merge pull request #4 from dimitarproynov/refactor_unit_tests

Remove the usage of "Environments" GitHub feature

Commit:
Update CHANGELOG.md for 1.10.1 release

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

dimitarproynov committed:
Merge pull request #5 from dimitarproynov/refactor_unit_tests

Merge refactor_unit_tests into main branch

dimitarproynov committed:
Merge pull request #6 from dimitarproynov/refactor_unit_tests

Sync Cluster EDRS properties with SDDC EDRS properties

dimitarproynov committed:
Merge pull request #7 from vmware/main

pull latest from original repo

Commit:
Run spectral scans as GitHub Actions

Testing done:
followed:

https://app.spectralops.io/sources/code/github-actions

Added SPECTRAL_DSN as a GitHub secret at repository level.

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Merge remote-tracking branch 'origin/main'

Commit:
Correct Spectral scan GitHub action name

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Run Cluster Create/Update/Delete tasks sequentially per SDDC

If two concurrent cluster tasks run per SDDC we get the following error:
com.vmware.skyscraper.common.InternalSkyscraperException: Concurrent update on <SDDC-ID>, column name {1}. Try operation again.

Use a per SDDC lock to force the Terraform provider to wait for the previous task completion.

Now that an almost full E2E suite is present, the pattern of the copy-pasted "wait until finished or error" code across the provider's resources can be extracted, as a helper function to avoid code duplication, safely.

Testing done:
make build
make test
golangci-lint run

Added a second cluster in the TestAccResourceVmcClusterZerocloud

ran successfully all acceptance tests:
make testacc TESTARGS="-run='TestAccResourceVmcSddcZerocloud|TestAccResourceVmcClusterZerocloud|TestAccResourceVmcSiteRecoveryZerocloud|TestAccResourceVmcSrmNodeZerocloud|TestAccDataSourceVmcCustomerSubnetsBasic|TestAccDataSourceVmcConnectedAccountsBasic|TestAccDataSourceVmcOrgBasic|TestAccDataSourceVmcSddcBasic' -parallel 4

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

fstrudel commented on vmc/resource_vmc_sddc.go:
```diff
@@ -787,26 +724,14 @@ func resourceSddcUpdate(d *schema.ResourceData, m interface{}) error {
 		return resource.RetryContext(context.Background(), d.Timeout(schema.TimeoutCreate), func() *resource.RetryError {
-			tasksClient := orgs.NewTasksClient(connectorWrapper)
-			task, err := tasksClient.Get(orgID, task.Id)
-			if err != nil || *task.Status == "FAILED" {
```
nit: microsoft -> Microsoft

fstrudel commented on vmc/sync_utils.go:
```diff
@@ -0,0 +1,100 @@
+/* Copyright 2022 VMware, Inc.
+   SPDX-License-Identifier: MPL-2.0 */
+
+package vmc
```
As a general Golang rule, avoid `util` and `helper` naming, see https://dave.cheney.net/2019/01/08/avoid-package-names-like-base-util-or-common

Sounds like this file provides tasks related methods?

dimitarproynov replied:
Extracted the task related functions to a separate package.

fstrudel commented on vmc/sync_utils_test.go:
```diff
@@ -0,0 +1,188 @@
+package vmc
```
always add the proper copyright

dimitarproynov replied:
Added copyright headers and updated all the stale ones.

fstrudel commented on vmc/sync_utils_test.go:
```diff
@@ -0,0 +1,188 @@
+	assert.True(t, lock1Obtained)
+}
+
+type AuthenticatorMock struct {
```
Those are not really mocks but stubs. 
You might want to generate the mocks actually to use proper mocks? See https://github.com/golang/mock

dimitarproynov replied:
Renamed to Stubs

Commit:
Refactor task related code

Task related code has been extracted in to a separate package.
The client connector related code has been moved to a separate package so that the Authenticator interface can be used in "task" package without a dependency loop to "vmc" package.
Constants have been extracted to a separate package as well to avoid dependency loops as well.

Testing done:
make build
make test
golangci-lint run

ran successfully all acceptance tests:
make testacc TESTARGS="-run='TestAccResourceVmcSddcZerocloud|TestAccResourceVmcClusterZerocloud|TestAccResourceVmcSiteRecoveryZerocloud|TestAccResourceVmcSrmNodeZerocloud|TestAccDataSourceVmcCustomerSubnetsBasic|TestAccDataSourceVmcConnectedAccountsBasic|TestAccDataSourceVmcOrgBasic|TestAccDataSourceVmcSddcBasic' -parallel 4

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>

Commit:
Include asset mapping for Spectral scan

As recommended by VMware Spectral integration team.

Testing done:
make build
make test
golangci-lint run

Signed-off-by: Dimitar Proynov <proynovd@vmware.com>
